frontend/src/shared/metadata/instance-type-dropdown.tsx

```diff
@@ -40,7 +43,98 @@ export type InstanceTypeSelectorProperties = {
         | 'ProductionVariantInstanceType';
 };
 
+type InstanceTypeMultiSelectorProperties = InstanceTypeProperties & {
+    selectedOptions: OptionDefinition[]; 
+};
+
+type InstanceTypeSelectorProperties = InstanceTypeProperties & {
+    selectedOption: OptionDefinition | null;
+    service?: string;
+};
+
+function getInstanceTypes (computeTypes: string[]) {
+    const categorizedTypes: Record<string, OptionDefinition[]> = {
+        Standard: [],
+        'Compute Optimized': [],
+        'Memory Optimized': [],
+        'Accelerated computing': [],
+    };
+
+    computeTypes.forEach(
+        (instanceType) => {
+            // Instance types returned from this call are in the format of "ml.<family>.size".
+            // We can categorize these instances based on the first letter of the
+            // <family> value.
+            let category = 'Accelerated computing';
+            switch (instanceType.split('.')[1][0]) {
+                case 't':
+                case 'm':
+                    category = 'Standard';
+                    break;
+                case 'c':
+                    category = 'Compute Optimized';
+                    break;
+                case 'r':
+                    category = 'Memory Optimized';
+                    break;
+            }
+            categorizedTypes[category].push({ value: instanceType, label: instanceType });
+        }
+    );
+    const groups: OptionGroup[] = [];
+    Object.keys(categorizedTypes).forEach((category) => {
+        if (categorizedTypes[category].length > 0) {
+            groups.push({ label: category, options: categorizedTypes[category] });
+        }
+    });
+    return groups;
+}
+
 export function InstanceTypeSelector (props: InstanceTypeSelectorProperties) {
+    const dispatch = useAppDispatch();
+    const applicationConfig: IAppConfiguration = useAppSelector(appConfig);
+    const computeTypes = useAppSelector(selectComputeTypes);
+    const [instanceOptions, setInstanceOptions] = useState([] as OptionDefinition[]);
```

Comment:
We should get out of the habit of using `useState` for things that can be computed at render (and use `useMemo` if there is a performance concern) so we aren't updating state in response to state changing. See [You Might Not Need an Effect](https://react.dev/learn/you-might-not-need-an-effect#caching-expensive-calculations).  

Something like the following...

```typescript
const instanceOptions = useMemo(() => {
  if (props.service) {
    // might need some safe defaults using `||` and using `?.` if things this chain are nullable
    return getInstanceTypes(applicationConfig.configuration.EnabledInstanceTypes[props.service]);
  } else {
    // might need some safe defaults using `||` and using `?.` if things this chain are nullable
    return getInstanceTypes(computeTypes.values.InstanceTypes[props.instanceTypeCategory || 'InstanceType']);
  }
}, [props, applicationConfig.configuration.EnabledInstanceTypes, computeTypes.values.InstanceTypes])

useEffect(() => {
  if (computeTypes.status === LoadingStatus.INITIAL) {
    dispatch(listComputeTypes());
  }
}, [dispatch, computeTypes.status])

Reply:
This is a good note. That being said, the React article I didn't really find helpful.

I find this Stackoverflow article better at explaining how useState / useEffect trigger unnecessary renders: https://stackoverflow.com/questions/56028913/usememo-vs-useeffect-usestate

Reply:
I appreciate you providing example code for this, makes my life a lot easier